File: LeetCode/Medium/0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        n = len(nums)
        memo = {}
        
        def dp(start, end):
            if (start, end) in memo:
                return memo[(start, end)]
            
            if start == end:
                return nums[start]
            
            memo[(start, end)] = max(nums[start] - dp(start + 1, end), nums[end] - dp(start, end - 1))
            
            return memo[(start, end)]
        
        return dp(0, n-1) >= 0
        # n = len(nums)
        # player_sum = 0
        # def solve(nums: List[int], start, end, player_sum, turn):
        #     if start > end:
        #         if player_sum >= 0:
        #             return True
        #         else:
        #             return False

        #     first = nums[start]
        #     last = nums[end]
        #     # player1 turn
        #     if turn == 0:
        #         return (solve(nums, start+1, end, player_sum + first, 1) or
        #                 solve(nums, start, end-1, player_sum + last, 1))
        #     # player2 turn
        #     elif turn == 1:
        #         return (solve(nums, start+1, end, player_sum - first, 0) and
        #                 solve(nums, start, end-1, player_sum - last, 0))

        # return solve(nums, 0, n-1, 0, 0)

File: LeetCode/Medium/0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution{
    public:
        int solve(vector<vector<int>>&dp, vector<int>& nums, int i, int j, bool turn){
            if(i>j)
                return 0;
            if(dp[i][j]!=-1e5)
                return dp[i][j];
            int ans = 0;
            if(turn){
                ans = max(nums[i]+solve(dp, nums, i+1, j, false), nums[j]+solve(dp, nums, i, j-1, false));
            }
            else
                ans = min(-nums[i]+solve(dp, nums, i+1, j, true), -nums[j]+solve(dp,nums,i,j-1,true));
            return dp[i][j]=ans;
        }
        bool PredictTheWinner(vector<int>& nums){
            int n = nums.size();
            vector<vector<int>>dp(n, vector<int>(n,-1e5));
            return solve(dp,nums,0,n-1,true) >= 0;
        }
};

// O(2^n)
// class Solution {
// public:
//     int solve(vector<int>& nums, int i, int j, bool turn){
//         if(i>j)
//             return 0;

//         // turn == 1 : player 1
//         if(turn){
//             return max(nums[i] + solve(nums,i+1,j,false), nums[j]+solve(nums,i, j-1, false));
//         }
//         // else : player 2
//         else
//             return min(solve(nums,i+1,j,true) - nums[i], solve(nums,i,j-1,true)-nums[j]);
//     }
//     bool PredictTheWinner(vector<int>& nums) {
//         int ans = solve(nums, 0, nums.size()-1, true);
//         return ans >= 0;
//     }
// };